### PR TITLE
Feature/modernization+hardening

### PR DESCRIPTION
**Pull request by SveinRoarKvale**
## Summary

Stabilizes discovery/object enumeration against flaky devices and mis-addressed replies: smarter object-list decoding, optional invoke/source correlation on the transaction layer, and tests.

## Changes

- **Object list / metadata (`objectlist.go`)** — Parse by property id (order-independent), distinguish indexed vs sequential object-list responses, tighter errors (device instance, property, types). Coerce list length and object type from common numeric variants (e.g. float/narrow int) where gateways mis-encode; object name remains string-only.
- **TSM (`tsm/transactions.go`)** — Optional `ExpectSource` + `SendFrom` so responses match the BACnet source after `SetLength` fills `Addr`.
- **Read paths** — `readprop` / `readmulti` set expected source where applicable; small `device.go` dispatch glue.
- **Tests** — `objectlist_test.go` (decode/corner cases), extended `tsm/transactions_test.go` (match/mismatch/fallback).

## Risk / rollout

Backward-compatible: correlation is opt-in; existing callers without `ExpectSource` behave as before.

## Verify

`go test ./... -race -short`

### CODE REVIEW

**Comment by SveinRoarKvale** [comment]
We should protect the main branch I think
I pushed wrongly first so it shows the history and changes slightly wrong
But what I tried to do was to allow for tighter control on packets src/dst addresses, and to not rely on index of objects in a given order, but decode and return the object type as well

Also general update on error logging and a refresh of dependencies

File: objectlist.go [hunk 1]
```diff
@@ -203,3 +185,151 @@ func (c *client) Objects(dev btypes.Device) (btypes.Device, error) {
...
 	return dev, nil
 }
+
+func extractObjectListLength(devID btypes.ObjectID, props []btypes.Property) (int, error) {
+	for _, p := range props {
+		if p.Type != btypes.PropObjectList || p.ArrayIndex != 0 {
+			continue
+		}
+		n, ok := unsignedFromPropertyData(p.Data)
+		if !ok {
```
Comment (afonso360) [inline objectlist.go]: Is this allowed by the spec? A list length arriving as a float?

File: objectlist.go [hunk 2]
```diff
@@ -203,3 +185,151 @@ func (c *client) Objects(dev btypes.Device) (btypes.Device, error) {
 	}
 	return dev, nil
 }
+
+func extractObjectListLength(devID btypes.ObjectID, props []btypes.Property) (int, error) {
+	for _, p := range props {
```
Comment (afonso360) [inline objectlist.go]: This seems to be unnecessary. It is only used in `ReadProperty`, which by spec can only return one property. We should enforce that constraint if possible, instead of silently allowing extra stuff.

Are we seeing this behavior from a device?

File: objectlist.go [hunk 3]
```diff
@@ -203,3 +185,151 @@ func (c *client) Objects(dev btypes.Device) (btypes.Device, error) {
...
+			switch v := p.Data.(type) {
+			case uint32:
+				// Many devices encode object type as uint32 even though model uses ObjectType.
+				objectTyp = btypes.ObjectType(v)
+				typeOK = true
+			case btypes.ObjectType:
+				objectTyp = v
+				typeOK = true
+			default:
+				if ot, ok := unsignedFromPropertyData(v); ok {
```
Comment (afonso360) [inline objectlist.go]: Is it legal for this to come in as float32/64, and are we seeing it a real device?

**Review by afonso360 — COMMENTED** [review]
Still going through this, but wanted to ask some clarifying questions first

File: objectlist.go (hunk 1 above)
Comment (SveinRoarKvale) [reply objectlist.go]: Not strictly no. It should be an unsigned int. I left it there to handle potential buggy gateway since one of the faults was related to this metrics and the value is decoded as interface{}. 
But agree - let's not allow it.

File: objectlist.go (hunk 2 above)
Comment (SveinRoarKvale) [reply objectlist.go]: it was a generalisation from my side having the possibility to iterate over multiple properties. 
I'll remove it

File: objectlist.go (hunk 3 above)
Comment (SveinRoarKvale) [reply objectlist.go]: No; it is not. 
It was one the faults we got yes, but that should be better handled now that we check property types
